// coding=utf-8
// Copyright 2022 Foivos Tsimpourlas.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// This files defines the specification of the reinforcement learning model.
syntax = "proto2";

package benchpress;

option go_package = "benchpresspb";
option java_multiple_files = true;
option java_outer_classname = "ReinforcementLearningProto";
option java_package = "com.benchpress";

import "deeplearning/benchpress/proto/model.proto";

message RLModel {
  oneof target_features {
    bool train_set = 1;
    bool random    = 2;
  }
  optional RLAgent agent          = 3;
  optional Model   language_model = 4;
}

message RLAgent {
  optional FeatureTokenizer feature_tokenizer         = 1;
  optional int32            num_epochs                = 2;
  optional int32            num_episodes              = 3;
  optional int32            steps_per_episode         = 4;
  optional int32            num_updates               = 5;
  optional float            gamma                     = 6;
  optional float            lam                       = 7;
  optional float            epsilon                   = 8;
  optional float            learning_rate_micros      = 9;
  optional float            value_loss_coefficient    = 10;
  optional float            entropy_coefficient       = 11;
  optional int32            batch_size                = 12;
  optional int32            action_temperature_micros = 13;
  optional int32            token_temperature_micros  = 14;
}

message FeatureTokenizer {
  // Max value that can be tokenized. Anything above that is UNK.
  optional int32 feature_max_value_token    = 1;
  // Threshold of values being singularly tokenized (not represented as range).
  optional int32 feature_singular_token_thr = 2;
  // Range width of tokenized values after singularity threshold
  optional int32 feature_token_range        = 3;
  // Sequence length of tokenized feature vectors.
  optional int32 feature_sequence_length    = 4;
}
